e2e_test.py

```diff
@@ -0,0 +1,100 @@
+from colbert.infra import ColBERTConfig, RunConfig, Run
+from colbert.data import Queries
+
+CKPT = "/dfs/scratch0/okhattab/share/2021/checkpoints/msmarco.psg.kldR2.nway64.ib__colbert-400000"
+STACKEXCHANGE_DATA_PATH = "/future/u/xmgui/testdata"
+EXPERIMENT_DIR = "/future/u/xmgui/ColBERT/experiments"
+
+
+def build_index(dataset_path):
+    nbits = 2  # encode each dimension with 2 bits
+    doc_maxlen = 300  # truncate passages at 300 tokens
+    collection = os.path.join(dataset_path, 'cs224u.collection.tsv')
+    index_name = f'e2etest.{nbits}bits.latest'
+    experiment = (f"e2etest.nbits={nbits}",)
+
+    with Run().context(RunConfig(nranks=4)):
+        config = ColBERTConfig(
+            doc_maxlen=doc_maxlen,
+            nbits=nbits,
+            root=EXPERIMENT_DIR,
+            experiment=experiment,
+        )
+        indexer = Indexer(CKPT, config=config)
+        indexer.index(name=index_name, collection=collection, overwrite=True)
+
+        config = ColBERTConfig(
+            root=EXPERIMENT_DIR,
+            experiment=experiment,
+        )
+        searcher = Searcher(
+            index=index_name,
+            config=config,
+        )
+
+        return searcher
+
+
+def search(searcher, dataset_path, query_type, k):
+    queries = os.path.join(
+        dataset_path, f"questions.{query_type}.tsv"
+    )
+    queries = Queries(path=queries)
+    ranking = searcher.search_all(queries, k=k)
+    output_path = ranking.save(f"e2etest.{query_type}.ranking.tsv")
+    return output_path
+
+
+def evaluate_dataset(dataset_path, query_type, k, rankings_path):
+    if not os.path.exists(rankings_path):
+        print(f"[query_type={query_type}] Success@{k}: ???")
+        return
+    rankings = defaultdict(list)
+    with open(rankings_path, "r") as f:
+        for line in f:
+            items = line.strip().split("\t")
+            qid, pid, rank = items[:3]
+
+            qid = int(qid)
+            pid = int(pid)  # post id
+            rank = int(rank)
+            rankings[qid].append(pid)
+            assert rank == len(rankings[qid])
+
+    success = 0
+    qas_path = os.path.join(dataset_path, f"qas.{query_type}.jsonl")
+
+    # check if the generated ranking's top k intersects with the provided answer
+    num_total_qids = 0
+    with jsonlines.open(qas_path, mode="r") as f:
+        for line in f:
+            qid = int(line["qid"])
+            num_total_qids += 1
+            if qid not in rankings:
+                print(f"WARNING: qid {qid} not found in {rankings_path}!", file=sys.stderr)
+                continue
+            answer_pids = set(line["answer_pids"])
+            if len(set(rankings[qid][:k]).intersection(answer_pids)) > 0:
+                success += 1
+    print(
+        f"[query_type={query_type}] "
+        f"Success@{k}: {success / num_total_qids * 100:.1f}"
+    )
+
+
+def main():
+    k = 5
+    searcher = build_index(STACKEXCHANGE_DATA_PATH)
+    for query_type in ["search", "forum"]:
```

Comment:
So the issue with evaluating the LoTTE queries on this new dataset is that the gold labels for the LoTTE queries correspond to LoTTE passages rather than passages in this new dataset - so basically we might end up retrieving good passages but there's no way to score them using success@k unless we manually select new gold labels. Maybe for now we can just validate that the scores for the top-k passages are all > some threshold rather than computing success@k.

Reply:
Also maybe instead of LoTTE we should switch to using the SQuAD dev sample queries used in the 224U notebook: https://github.com/cgpotts/cs224u/blob/master/hw_openqa.ipynb

Reply:
So since SQuAD is an open-domain question answering dataset each data point comes with a question and answer (as opposed to a question and a list of gold passages) - so then we can  compute success without needing gold passage labels by just assuming a passage is positive if it contains any of the answer's substrings (see the `evaluate_retrieval_example` and `success_at_k`) functions in the notebook

Reply:
thanks for the pointers, updated